lints/lint_ext_tor_service_descriptor_hash_invalid.go

```diff
@@ -0,0 +1,318 @@
+ * Descriptor Hash extension has the following format:
+ *
+ * cabf-TorServiceDescriptor OBJECT IDENTIFIER ::= { 2.23.140.1.31 }
+ *
+ * TorServiceDescriptorSyntax ::=
+ * SEQUENCE ( 1..MAX ) of TorServiceDescriptorHash
+ *
+ * TorServiceDescriptorHash:: = SEQUENCE {
+ *   onionURI              UTF8String
+ *   algorithm             AlgorithmIdentifier
+ *   subjectPublicKeyHash  BIT STRING
+ * }
+ *
+ * Where the AlgorithmIdentifier is a hashing algorithm (defined in RFC 6234)
+ * performed over the DER-encoding of an ASN.1 SubjectPublicKey of the .onion
+ * service and SubjectPublicKeyHash is the hash output.
+ */
+
+// TorServiceDescriptorHash holds an onion URI (e.g. `http://<whatever>.onion`),
+// a pkix.AlgorithmIdentifier, a slice of bytes holding a hash of the onion
+// site's public key created with the specified hash algorithm, and the number
+// of hash digest bits in the slice.
+type TorServiceDescriptorHash struct {
+	Onion    string
+	Alg      pkix.AlgorithmIdentifier
+	Hash     []byte
+	HashBits int
+}
+
+// Valid returns nil if the the TorServiceDescriptorHash specifies a valid hash
+// algorithm OID and the correct corresponding HashBits. Otherwise an error is
+// returned describing the problem.
+func (t TorServiceDescriptorHash) Valid() error {
+	// If the length is <= 0 its definitely not valid.
+	if t.HashBits <= 0 {
+		return errors.New("invalid TorServiceDescriptorHash subjectPublicKeyHash, " +
+			"bit length is <= 0")
+	}
+
+	// Check that the specified algorithm OID is an allowed hash algorithm OID.
+	allowedHashAlgs := []asn1.ObjectIdentifier{
+		util.SHA256OID,
+		util.SHA384OID,
+		util.SHA512OID,
+	}
+	goodHashAlg := util.SliceContainsOID(allowedHashAlgs, t.Alg.Algorithm)
+	if !goodHashAlg {
+		return fmt.Errorf("invalid TorServiceDescriptorHash algorithm %q",
+			t.Alg.Algorithm)
+	}
+
+	// Check that the number of bits in the hash field match the algorithm
+	// specified.
+	if t.Alg.Algorithm.Equal(util.SHA256OID) && t.HashBits != 256 {
+		return fmt.Errorf("invalid TorServiceDescriptorHash subjectPublicKeyHash, "+
+			"alg is SHA256 but bit length is %d not %d",
+			t.HashBits, 256)
+	} else if t.Alg.Algorithm.Equal(util.SHA384OID) && t.HashBits != 384 {
+		return fmt.Errorf("invalid TorServiceDescriptorHash subjectPublicKeyHash, "+
+			"alg is SHA384 but bit length is %d not %d",
+			t.HashBits, 384)
+	} else if t.Alg.Algorithm.Equal(util.SHA512OID) && t.HashBits != 512 {
+		return fmt.Errorf("invalid TorServiceDescriptorHash subjectPublicKeyHash, "+
+			"alg is SHA512 but bit length is %d not %d",
+			t.HashBits, 512)
+	}
+	return nil
+}
+
+// parseTorServiceDescriptorHash unmarshals a SEQUENCE from the provided data
+// and parses a TorServiceDescriptorHash using the data contained in the
+// sequence. The TorServiceDescriptorHash object and the remaining data are
+// returned if no error occurs.
+func parseTorServiceDescriptorHash(data []byte) (*TorServiceDescriptorHash, []byte, error) {
```

Comment:
I suspect we want this parsing to live in ZCrypto (https://github.com/zmap/zcrypto) and be an exposed part of a certificate rather than parse out here as part of the lint. That way when we parse certificates in other places, we have this data available. Thoughts?

Reply:
That makes sense to me :+1: I'll take a crack at pulling out a separate zcrypto PR.

Reply:
I haven't had a chance to start on dissecting this PR but I think my rough plan of attack will be:

1) Closing this PR.
2) Opening a new PR against `zlint` that adds `lint_onion_subject_validity_time_too_large.go` and `lint_san_dns_name_onion_not_ev_cert.go` since these can land right away.
3) Opening a PR against `zcrypto` to add the `TorServiceDescriptorHash` parsing code
4) Working through review of that branch until it lands.
5) Opening a final PR against `zlint` that adds an updated `lint_ext_tor_service_descriptor_hash_invalid.go` that lints without any ASN.1 parsing logic.

@zakird Is it fair to assume that since `zlint` doesn't pin/vendor its dependencies that writing a lint that only builds against `zcrypto` master is copacetic?